packages/machine-extractor/src/extractAction.ts

```diff
@@ -351,24 +272,47 @@ export const getActionCreatorName = (actionNode: ActionNode) =>
     ? actionNode.node.callee.name
     : undefined;
 
-function extractEventObject(
-  eventObject: t.ObjectExpression,
+const isNull = (node: t.Node | null): node is null => t.isNullLiteral(node);
+
+export function extractArrayRecursively(
+  array: t.ArrayExpression,
   fileContent: string,
-):
-  | { type: 'expression'; value: string }
-  | {
-      type: 'object';
-      value: Record<string, JsonEntry>;
-    } {
-  if (!eventObject.properties.every((prop) => t.isObjectProperty(prop))) {
-    return {
-      type: 'expression',
-      value: fileContent.slice(eventObject.start!, eventObject.end!),
-    };
-  }
+): JsonItem[] {
+  const extracted: JsonItem[] = [];
+  array.elements.forEach((elem) => {
+    if (isNull(elem)) {
+      extracted.push(null);
+    } else if (t.isArrayExpression(elem)) {
+      extracted.push(extractArrayRecursively(elem, fileContent));
+    } else if (t.isObjectExpression(elem)) {
+      extracted.push(extractObjectRecursively(elem, fileContent));
+    } else {
+      if (t.isLiteral(elem)) {
+        if (
+          t.isRegExpLiteral(elem) ||
+          isTemplateLiteralWithExpressions(elem) ||
+          t.isNullLiteral(elem)
+        ) {
+          extracted.push(fileContent.slice(elem.start!, elem.end!));
+        } else if (t.isTemplateLiteral(elem)) {
+          extracted.push(
+            elem.quasis[0].value.cooked ?? elem.quasis[0].value.raw,
+          );
+        } else {
+          extracted.push(elem.value);
+        }
+      }
+    }
+  });
+  return extracted;
+}
 
-  const extracted: Record<string, JsonEntry> = {};
-  eventObject.properties.forEach((prop) => {
+export function extractObjectRecursively(
```

Comment:
There is a good chunk of duplication between `extractObjectRecursively` (btw. that name implies that there is some non-recursive variant somewhere and I don't think that's the case) and `extractArrayRecursively` (and some other places, such as what happens for `assign`).

You could have `extractValue` that would handle all of those cases in the same way (for `assign` u'd call it separately for each property value that u don't want to extract as an expression)